icepyx/core/types.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from typing import Literal, TypedDict, Union
+
+from typing_extensions import NotRequired
+
+IcesatProductShortName = Literal[
+    "ATL01",
+    "ATL02",
+    "ATL03",
+    "ATL04",
+    "ATL06",
+    "ATL07",
+    "ATL07QL",
+    "ATL08",
+    "ATL09",
+    "ATL09QL",
+    "ATL10",
+    "ATL11",
+    "ATL12",
+    "ATL13",
+    "ATL14",
+    "ATL15",
+    "ATL16",
+    "ATL17",
+    "ATL19",
+    "ATL20",
+    "ATL21",
+    "ATL23",
+]
+
+CMRParamsBase = TypedDict(
+    "CMRParamsBase",
+    {
+        "temporal": NotRequired[str],
+        "options[readable_granule_name][pattern]": NotRequired[str],
+        "options[spatial][or]": NotRequired[str],
+        "readable_granule_name[]": NotRequired[str],
+    },
+)
+
+
+class CMRParamsWithBbox(CMRParamsBase):
+    bounding_box: str
+
+
+class CMRParamsWithPolygon(CMRParamsBase):
+    polygon: str
+
+
+CMRParams = Union[CMRParamsWithBbox, CMRParamsWithPolygon]
+
+
+class EGISpecificParamsBase(TypedDict):
+    """Common parameters for searching, ordering, or downloading from EGI.
+
+    See: https://wiki.earthdata.nasa.gov/display/SDPSDOCS/EGI+Programmatic+Access+Documentation
+
+    EGI shares parameters with CMR, so this data is used in conjunction with CMRParams
+    to build EGI requests.
+
+    TODO: Validate more strongly (with Pydantic and its annotated types?
+    https://docs.pydantic.dev/latest/concepts/types/#composing-types-via-annotated):
+
+    * version is 3 digits
+    * 0 < page_size <= 2000
+    """
+
+    short_name: IcesatProductShortName  # alias: "product"
+    version: str
+    page_size: int  # default 2000
+    page_num: int  # default 0
+
+
+class EGISpecificParamsSearch(EGISpecificParamsBase):
```

Comment:
It's pedantic and not worth spending time massaging since Harmony parameter names are better aligned with CMR, but technically these are CMR required search parameters/EGI required download parameters (but not EGI specific parameters). Functionally it doesn't matter, but see your statements about "naming" as a top challenge in Python. 😆

Reply:
Good call - would you suggest replacing the word "Specific" with "Required" in these names?

Reply:
That's great!

POI for future naming: we're currently in some water-muddying induced by which parameters are required in a given situation. The current list of EGISpecificParamsBase is actually required parameters for a CMR search AND EGI download (and there's no such thing as an EGI Search - EGI is only hit for order/download). But again, neither here nor there to make forward progress since it will all change anyway and there's better streamlining of parameter names across CMR + Harmony.

Reply:
6a4acd9